File: orchestrator-bridge/src/main/kotlin/com/orchestrator/mcp/bridge/local/ConfigWatcher.kt
package com.orchestrator.mcp.bridge.local

import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import org.slf4j.LoggerFactory
import java.nio.file.*

/**
 * Watches mcp-servers.json for changes using Java NIO WatchService.
 * Emits config change events with debounce.
 */
class ConfigWatcher(
    private val configPath: String,
    private val onChanged: (suspend (Map<String, ServerConfig>) -> Unit)? = null,
    private val debounceMs: Long = 1000,
) {
    private val logger = LoggerFactory.getLogger(ConfigWatcher::class.java)
    private var watchJob: Job? = null
    private val json = Json { ignoreUnknownKeys = true }

    val path: String get() = configPath

    /** Load and parse config from disk. Returns empty map on error. */
    fun loadConfig(): Map<String, ServerConfig> {
        return try {
            val raw = Path.of(configPath).toFile().readText()
            parseConfig(raw)
        } catch (e: Exception) {
            logger.error("Failed to load {}: {}", configPath, e.message)
            emptyMap()
        }
    }

    /** Start watching the config file for changes. */
    fun start(scope: CoroutineScope) {
        val file = Path.of(configPath)
        if (!file.toFile().exists()) {
            logger.warn("Config not found: {}", configPath)
            return
        }
        logger.info("Watching: {}", configPath)
        watchJob = scope.launch(Dispatchers.IO) { watchLoop(file) }
    }

    /** Stop watching. */
    fun stop() {
        watchJob?.cancel()
        watchJob = null
    }

    private suspend fun watchLoop(file: Path) {
        val dir = file.parent ?: return
        val fileName = file.fileName.toString()

        try {
            val watchService = dir.fileSystem.newWatchService()
            dir.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY)

            while (currentCoroutineContext().isActive) {
                val key = withContext(Dispatchers.IO) { watchService.take() }
                var changed = false
                for (event in key.pollEvents()) {
                    val ctx = event.context() as? Path
                    if (ctx?.toString() == fileName) changed = true
                }
                key.reset()

                if (changed) {
                    delay(debounceMs)
                    logger.info("Config changed, reloading...")
                    val config = loadConfig()
                    onChanged?.invoke(config)
                }
            }
            watchService.close()
        } catch (_: CancellationException) {
            // Normal shutdown
        } catch (e: Exception) {
            logger.error("Watch error: {}", e.message)
        }
    }

    private fun parseConfig(raw: String): Map<String, ServerConfig> {
        val root = json.parseToJsonElement(raw).jsonObject
        val servers = root["mcpServers"]?.jsonObject ?: return emptyMap()

        return servers.mapNotNull { (name, element) ->
            try {
                val obj = element.jsonObject
                val config = ServerConfig(
                    command = obj["command"]!!.jsonPrimitive.content,
                    args = obj["args"]?.jsonArray?.map { it.jsonPrimitive.content } ?: emptyList(),
                    env = obj["env"]?.jsonObject?.mapValues { it.value.jsonPrimitive.content } ?: emptyMap(),
                    timeout = obj["timeout"]?.jsonPrimitive?.longOrNull ?: 30_000,
                    maxRetries = obj["maxRetries"]?.jsonPrimitive?.intOrNull ?: 3,
                    disabled = obj["disabled"]?.jsonPrimitive?.booleanOrNull ?: false,
                )
                name to config
            } catch (e: Exception) {
                logger.warn("Invalid config for server '{}': {}", name, e.message)
                null
            }
        }.toMap()
    }

    companion object {
        /** Resolve config path: CLI arg → CWD → home directory. */
        fun resolveConfigPath(cliPath: String? = null): String {
            if (cliPath != null) {
                val abs = Path.of(cliPath).toAbsolutePath()
                if (abs.toFile().exists()) return abs.toString()
            }
            val cwdPath = Path.of("mcp-servers.json").toAbsolutePath()
            if (cwdPath.toFile().exists()) return cwdPath.toString()

            val home = System.getenv("HOME") ?: System.getenv("USERPROFILE") ?: "."
            val homePath = Path.of(home, ".mcp-bridge", "mcp-servers.json")
            if (homePath.toFile().exists()) return homePath.toString()

            return cwdPath.toString()
        }
    }
}

File: mcp-client-bridge/src/local/config-watcher.ts
/**
 * Watches mcp-servers.json for changes with debounce.
 * Emits 'configChanged' when file is modified.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { EventEmitter } from 'node:events';
import { ServerConfig } from './server-process.js';

export interface McpServersConfig {
  mcpServers: Record<string, ServerConfig>;
}

export class ConfigWatcher extends EventEmitter {
  private watcher: fs.FSWatcher | null = null;
  private debounceTimer: ReturnType<typeof setTimeout> | null = null;
  private readonly debounceMs = 1000;

  constructor(private configPath: string) {
    super();
  }

  get path(): string { return this.configPath; }

  /** Load and parse config from disk. Returns empty config on error. */
  loadConfig(): McpServersConfig {
    try {
      const raw = fs.readFileSync(this.configPath, 'utf-8');
      const parsed = JSON.parse(raw) as McpServersConfig;
      if (!parsed.mcpServers || typeof parsed.mcpServers !== 'object') {
        console.error(`[config-watcher] Invalid config: missing mcpServers`);
        return { mcpServers: {} };
      }
      return parsed;
    } catch (err) {
      console.error(`[config-watcher] Failed to load ${this.configPath}:`, err);
      return { mcpServers: {} };
    }
  }

  /** Start watching the config file for changes. */
  start(): void {
    if (!fs.existsSync(this.configPath)) {
      console.error(`[config-watcher] Config not found: ${this.configPath}`);
      return;
    }
    try {
      this.watcher = fs.watch(this.configPath, () => this.onFileChange());
      console.error(`[config-watcher] Watching: ${this.configPath}`);
    } catch (err) {
      console.error(`[config-watcher] Watch failed:`, err);
    }
  }

  stop(): void {
    if (this.debounceTimer) clearTimeout(this.debounceTimer);
    this.watcher?.close();
    this.watcher = null;
  }

  private onFileChange(): void {
    if (this.debounceTimer) clearTimeout(this.debounceTimer);
    this.debounceTimer = setTimeout(() => {
      console.error('[config-watcher] Config changed, reloading...');
      this.emit('configChanged', this.loadConfig());
    }, this.debounceMs);
  }

  /** Resolve config path: CLI arg → CWD → home directory. */
  static resolveConfigPath(cliPath?: string): string {
    if (cliPath) {
      const abs = path.isAbsolute(cliPath) ? cliPath : path.resolve(cliPath);
      if (fs.existsSync(abs)) return abs;
    }
    const cwdPath = path.resolve(process.cwd(), 'mcp-servers.json');
    if (fs.existsSync(cwdPath)) return cwdPath;

    const homePath = path.resolve(
      process.env.HOME ?? process.env.USERPROFILE ?? '.',
      '.mcp-bridge',
      'mcp-servers.json',
    );
    if (fs.existsSync(homePath)) return homePath;

    // Return CWD path even if not found (will be created or error later)
    return cwdPath;
  }
}
